plutus-tx/src/PlutusTx/Data/AssocMap.hs

```diff
@@ -71,6 +90,18 @@ instance P.UnsafeFromData (Map k a) where
   {-# INLINABLE unsafeFromBuiltinData #-}
   unsafeFromBuiltinData = Map . BI.unsafeDataAsMap
 
+{-# INLINEABLE size #-}
+size :: forall k a. Map k a -> Integer
+size (Map m) = go m
+  where
+    go l =
+      P.matchList
+        l
+        (\() -> 0)
+        ( \_ tl ->
+            1 + go tl
```

Comment:
Have you tried using an accumulator to make it tail recursive? I wonder if it affects the execution costs.

Somethiing like

size (Map m) = go m 0
   where
       go l acc = 
             P.matchList l (\() -> acc)  ( \_ tl -> go tl (acc `addInteger` 1))
```